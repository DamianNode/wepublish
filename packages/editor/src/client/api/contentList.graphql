query ContentList(
  $type: contentTypeEnum!
  $context: contentContextEnum
  $filter: String
  $after: ID
  $first: Int
) {
  content {
    _all {
      list(type: $type, context: $context, first: $first, after: $after, filter: {title: $filter}) {
        nodes {
          peer {
            id
            name
          }
          content {
            ...ContentListRef
          }
        }
        pageInfo {
          startCursor
          endCursor
          hasNextPage
          hasPreviousPage
        }
        totalCount
      }
    }
  }
}

query ContentGet($id: ID!) {
  content {
    _all {
      read(id: $id) {
        ...ContentListRef
      }
    }
  }
}

fragment ContentListRef on ContentModelSummary {
  id
  title
  shared
  contentType
  revision
  state
  createdAt
  modifiedAt
  publicationDate
  dePublicationDate
}

mutation PublishContent(
  $id: ID!
  $publishAt: DateTime!
  $publishedAt: DateTime!
  $updatedAt: DateTime!
) {
  content {
    _all {
      publish(id: $id, publishAt: $publishAt, publishedAt: $publishedAt, updatedAt: $updatedAt) {
        id
        pending {
          publishAt
        }
        published {
          publishedAt
        }
      }
    }
  }
}

mutation UnpublishContent($id: ID!) {
  content {
    _all {
      unpublish(id: $id) {
        id
        pending {
          publishAt
        }
        published {
          publishedAt
        }
      }
    }
  }
}

mutation DeleteContent($id: ID!) {
  content {
    _all {
      delete(id: $id)
    }
  }
}
