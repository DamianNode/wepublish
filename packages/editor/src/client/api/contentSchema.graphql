query Config {
  config {
    content {
      id
      identifier
      namePlural
      nameSingular
      schema
    }
    languages {
      defaultLanguageId
      languages {
        id
        tag
        description
      }
    }
  }
}
