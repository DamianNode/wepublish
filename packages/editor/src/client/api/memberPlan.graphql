fragment MemberPlanRef on MemberPlan {
    id
    label
    isActive
    image {
        ...ImageRef
    }
}

fragment FullMemberPlan on MemberPlan {
    description
    pricePerMonthMinimum
    pricePerMonthMaximum
    availablePaymentMethods {
        paymentMethod {
            ...FullPaymentMethod
        }
        paymentPeriodicity {
            id
            checked
        }
        minimumDurationMonths
        forceAutoRenewal
    }
    ...MemberPlanRef
}

query MemberPlanList($filter: String, $after: ID, $before: ID, $first: Int, $last: Int) {
    memberPlans(filter: {label: $filter}, after: $after, before: $before, first: $first, last: $last) {
        nodes {
            ...FullMemberPlan
        }

        pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
        }

        totalCount
    }
}

query MemberPlan($id: ID!) {
    memberPlan(id: $id) {
        ...FullMemberPlan
    }
}

mutation CreateMemberPlan($input: MemberPlanInput!) {
    createMemberPlan(input: $input) {
        ...FullMemberPlan
    }
}

mutation UpdateMemberPlan($id: ID!, $input: MemberPlanInput!) {
    updateMemberPlan(id: $id, input: $input) {
        ...FullMemberPlan
    }
}

mutation DeleteMemberPlan($id: ID!) {
    deleteMemberPlan(id: $id)
}
