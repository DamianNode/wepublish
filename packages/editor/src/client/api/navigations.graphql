fragment FullNavigation on Navigation {
  id
  name
  key
  links {
    __typename

    ... on PageNavigationLink {
      label
      page {
        ...PageRef
      }
    }
    ... on ArticleNavigationLink {
      label
      article {
        ...ArticleRef
      }
    }
    ... on ExternalNavigationLink {
      label
      url
    }
  }
}

query NavigationsList {
  navigations {
    ...FullNavigation
  }
}

query Navigation($id: ID!) {
  navigation(id: $id) {
    ...FullNavigation
  }
}

mutation CreateNavigation($input: NavigationInput!) {
  createNavigation(input: $input) {
    ...FullNavigation
  }
}

mutation UpdateNavigation($id: ID!, $input: NavigationInput!) {
  updateNavigation(id: $id, input: $input) {
    ...FullNavigation
  }
}

mutation DeleteNavigation($id: ID!) {
  deleteNavigation(id: $id) {
    ...FullNavigation
  }
}
