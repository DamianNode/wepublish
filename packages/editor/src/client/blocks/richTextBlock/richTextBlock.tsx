import React, {memo, useMemo, useState, useEffect} from 'react'
import {useTranslation} from 'react-i18next'
import {createEditor, Location, Node as SlateNode, Transforms} from 'slate'
import {withHistory} from 'slate-history'
import {withReact, ReactEditor, Editable, Slate} from 'slate-react'
import {BlockProps} from '../../atoms/blockList'
import {EmojiPicker} from '../../atoms/emojiPicker'
import {
  Toolbar,
  ToolbarDivider,
  H1Icon,
  H2Icon,
  H3Icon,
  SubMenuButton,
  SubMenuModalButton
} from '../../atoms/toolbar'
import {RichTextBlockValue} from '../types'
import {FormatButton, FormatIconButton, EditorSubMenuButton} from './toolbar/buttons'
import {renderElement, renderLeaf} from './editor/render'
import {BlockFormat, TextFormat} from './editor/formats'
import {withRichText, withTable} from './editor/plugins'
import {withNormalizeNode} from './editor/normalizing'
import {TableMenu} from './toolbar/tableMenu'
import {WepublishEditor} from './editor/wepublishEditor'
import {LinkMenu} from './toolbar/linkMenu'
import {RefMenu} from './toolbar/refMenu'
import {ContentModelSchemaFieldRefTypeMap} from '../../interfaces/referenceType'

export interface RichTextConfig {
  h1?: boolean
  h2?: boolean
  h3?: boolean
  italic?: boolean
  bold?: boolean
  underline?: boolean
  strikethrough?: boolean
  superscript?: boolean
  subscript?: boolean
  table?: boolean
  emoji?: boolean
  unorderedList?: boolean
  orderedList?: boolean
  url?: boolean
  ref?: ContentModelSchemaFieldRefTypeMap
}

export interface RichTextBlockProps extends BlockProps<RichTextBlockValue> {
  displayOnly?: boolean
  showCharCount?: boolean
  displayOneLine?: boolean
  config?: RichTextConfig
}

export const RichTextBlock = memo(function RichTextBlock({
  value,
  autofocus,
  disabled,
  onChange,
  displayOnly = false,
  showCharCount = false,
  displayOneLine = false,
  config = {
    h1: true,
    h2: true,
    h3: true,
    italic: true,
    bold: true,
    underline: true,
    strikethrough: true,
    superscript: true,
    subscript: true,
    table: true,
    emoji: true,
    unorderedList: true,
    orderedList: true,
    url: true,
    ref: undefined
  }
}: RichTextBlockProps) {
  const editor = useMemo(
    () => withNormalizeNode(withTable(withRichText(withHistory(withReact(createEditor()))))),
    []
  )
  const [hasFocus, setFocus] = useState(false)
  const [location, setLocation] = useState<Location | null>(null)

  const [charCount, setCharCount] = useState(0)

  useEffect(() => {
    setCharCount(WepublishEditor.calculateEditorCharCount(editor))
  }, [editor.children])

  const {t} = useTranslation()

  useEffect(() => {
    if (autofocus) {
      ReactEditor.focus(editor)
    }
  }, [])

  const focusAtPreviousLocation = (location: Location) => {
    Transforms.select(editor, location)
    ReactEditor.focus(editor)
  }

  const activateHotkey = (e: React.KeyboardEvent<HTMLDivElement>) => {
    switch (e.key.toLowerCase()) {
      case 'b': {
        e.preventDefault()
        WepublishEditor.toggleFormat(editor, TextFormat.Bold)
        break
      }
      case 'i': {
        e.preventDefault()
        WepublishEditor.toggleFormat(editor, TextFormat.Italic)
        break
      }
      case 'u': {
        e.preventDefault()
        WepublishEditor.toggleFormat(editor, TextFormat.Underline)
        break
      }
      case 'x': {
        if (e.getModifierState('Shift')) {
          e.preventDefault()
          WepublishEditor.toggleFormat(editor, TextFormat.Strikethrough)
        }
        break
      }
    }
  }

  return (
    <Slate
      editor={editor}
      value={value}
      onChange={(newValue: SlateNode[]) => {
        setFocus(ReactEditor.isFocused(editor))
        if (value !== newValue) {
          onChange(newValue)
        }
      }}>
      {!displayOnly && (
        <>
          <Toolbar
            fadeOut={!hasFocus}
            onMouseDown={() => {
              // e.preventDefault()
              if (!hasFocus && location) focusAtPreviousLocation(location)
            }}>
            {config.h1 && (
              <FormatButton format={BlockFormat.H1}>
                <H1Icon />
              </FormatButton>
            )}
            {config.h2 && (
              <FormatButton format={BlockFormat.H2}>
                <H2Icon />
              </FormatButton>
            )}
            {config.h3 && (
              <FormatButton format={BlockFormat.H3}>
                <H3Icon />
              </FormatButton>
            )}
            {(config.h1 || config.h2 || config.h3) && <ToolbarDivider />}

            {config.unorderedList && (
              <FormatIconButton icon="list-ul" format={BlockFormat.UnorderedList} />
            )}
            {config.orderedList && (
              <FormatIconButton icon="list-ol" format={BlockFormat.OrderedList} />
            )}

            {(config.orderedList || config.unorderedList) && <ToolbarDivider />}

            {config.table && (
              <>
                <EditorSubMenuButton icon="table" editorHasFocus={hasFocus}>
                  <TableMenu />
                </EditorSubMenuButton>
                <ToolbarDivider />
              </>
            )}

            {config.bold && <FormatIconButton icon="bold" format={TextFormat.Bold} />}
            {config.italic && <FormatIconButton icon="italic" format={TextFormat.Italic} />}
            {config.underline && (
              <FormatIconButton icon="underline" format={TextFormat.Underline} />
            )}
            {config.strikethrough && (
              <FormatIconButton icon="strikethrough" format={TextFormat.Strikethrough} />
            )}
            {config.superscript && (
              <FormatIconButton icon="superscript" format={TextFormat.Superscript} />
            )}
            {config.subscript && (
              <FormatIconButton icon="subscript" format={TextFormat.Subscript} />
            )}
            {(config.bold ||
              config.italic ||
              config.underline ||
              config.strikethrough ||
              config.superscript ||
              config.subscript) && <ToolbarDivider />}

            {config?.url && (
              <>
                <SubMenuButton icon="link">
                  <LinkMenu />
                </SubMenuButton>
                <ToolbarDivider />
              </>
            )}

            {config?.ref && (
              <>
                <SubMenuModalButton icon="link">
                  <RefMenu types={config?.ref} />
                </SubMenuModalButton>
                <ToolbarDivider />
              </>
            )}

            {config.emoji && (
              <SubMenuButton icon="smile-o">
                <EmojiPicker setEmoji={emoji => editor.insertText(emoji)} />
              </SubMenuButton>
            )}
          </Toolbar>
          {WepublishEditor.isEmpty(editor) && ( // Alternative placeholder
            <div onClick={() => ReactEditor.focus(editor)} style={{color: '#cad5e4'}}>
              {t('blocks.richText.startWriting')}
            </div>
          )}
        </>
      )}
      <Editable
        style={
          displayOneLine
            ? {
                whiteSpace: 'nowrap',
                overflow: 'hidden',
                textOverflow: 'ellipsis'
              }
            : undefined
        }
        readOnly={disabled || displayOnly}
        // placeholder={t('blocks.richText.startWriting')}  # causes focusing problems on firefox !
        renderElement={renderElement}
        renderLeaf={renderLeaf}
        onBlur={() => {
          setLocation(editor.selection)
        }}
        onKeyDown={e => {
          if (e.ctrlKey || e.metaKey) activateHotkey(e)
        }}
      />
      {showCharCount && (
        <p style={{textAlign: 'right'}}>{t('blocks.richText.charCount', {charCount: charCount})}</p>
      )}
    </Slate>
  )
})

export function createDefaultValue(): RichTextBlockValue {
  return WepublishEditor.createDefaultValue()
}
