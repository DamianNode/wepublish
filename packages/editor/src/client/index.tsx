import {initClient} from './client'
import {ExtensionConfig} from './interfaces/extensionConfig'

import 'rsuite/lib/styles/index.less'
import './global.less'
import './atoms/emojiPicker.less'
import './atoms/toolbar.less'
import './blocks/richTextBlock/toolbar/tableMenu.less'

const config: ExtensionConfig = {}
if (document.readyState !== 'loading') {
  initClient(config)
} else {
  document.addEventListener('DOMContentLoaded', () => {
    initClient(config)
  })
}
