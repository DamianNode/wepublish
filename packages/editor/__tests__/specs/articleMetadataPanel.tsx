import React from 'react'
import {MockedProvider as MockedProviderBase} from '@apollo/client/testing'
import {ArticleMetadataPanel} from '../../src/client/panel/articleMetadataPanel'
import {AuthorListDocument} from '../../src/client/api'
import {mount} from 'enzyme'

import {UIProvider} from '@karma.run/ui'
import {act} from 'react-dom/test-utils'
import {updateWrapper} from '../utils'
import * as fela from 'fela'
import {createDefaultValue} from '../../src/client/blocks/richTextBlock'

const MockedProvider = MockedProviderBase as any

const styleRenderer: fela.IRenderer = {
  renderRule: jest.fn(),
  renderKeyframe: jest.fn(),
  renderFont: jest.fn(),
  renderStatic: jest.fn(),
  renderToString: jest.fn(),
  subscribe: jest.fn(),
  clear: jest.fn()
}
const article = {
  slug: 'the-properties-of-moonstone-and-its-uses',
  preTitle: 'Moonstone Uses',
  title: 'The Properties of Moonstone And Its Uses',
  lead: '',
  authors: [
    {
      id: 'authorId2',
      name: 'Hermione Granger'
    }
  ],
  tags: ['moonstone', 'potions'],
  properties: [
    {
      key: 'abc',
      value: 'def',
      public: false
    }
  ],
  image: undefined,
  shared: false,
  breaking: true
}

const mocks = [
  {
    request: {
      query: AuthorListDocument,
      variables: {
        first: 10,
        filter: undefined
      }
    },
    result: () => ({
      data: {
        authors: {
          nodes: [
            {
              __typename: 'Author',
              id: 'authorId2',
              name: 'Hermione Granger',
              slug: 'hermione-granger',
              //url: 'www.hg.ch/',
              links: [{title: 'Link Title', url: 'www.link-url.ch/'}],
              //bio: createDefaultValue(),
              image: undefined
            }
          ],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          },
          totalCount: 1
        }
      }
    })
  }
]

describe('Article Metadata Panel', () => {
  test('should render', async () => {
    const wrapper = mount(
      <UIProvider styleRenderer={styleRenderer} rootElementID={'fskr'}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <ArticleMetadataPanel value={article} />
        </MockedProvider>
      </UIProvider>
    )
    await updateWrapper(wrapper, 100)

    const panel = wrapper.find('ArticleMetadataPanel')
    expect(panel).toMatchSnapshot()
  })
})
