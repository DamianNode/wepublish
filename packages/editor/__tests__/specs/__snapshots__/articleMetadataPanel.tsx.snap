// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Metadata Panel should render 1`] = `
<ArticleMetadataPanel
  value={
    Object {
      "authors": Array [
        Object {
          "id": "authorId2",
          "name": "Hermione Granger",
        },
      ],
      "breaking": true,
      "image": undefined,
      "lead": "",
      "preTitle": "Moonstone Uses",
      "properties": Array [
        Object {
          "key": "abc",
          "public": false,
          "value": "def",
        },
      ],
      "shared": false,
      "slug": "the-properties-of-moonstone-and-its-uses",
      "tags": Array [
        "moonstone",
        "potions",
      ],
      "title": "The Properties of Moonstone And Its Uses",
    }
  }
>
  <Styled(div)>
    <div>
      <PanelHeader
        leftChildren={
          <ForwardRef(NavigationButton)
            icon={[Function]}
            label="articleEditor.panels.close"
            onClick={[Function]}
          />
        }
        title="articleEditor.panels.metadata"
      >
        <Styled(div)>
          <div>
            <ForwardRef(NavigationButton)
              icon={[Function]}
              label="articleEditor.panels.close"
              onClick={[Function]}
            >
              <Styled(button)
                onClick={[Function]}
                styleProps={
                  Object {
                    "active": undefined,
                    "disabled": undefined,
                  }
                }
              >
                <button
                  onClick={[Function]}
                >
                  <Icon
                    element={[Function]}
                    scale="2em"
                  >
                    <MaterialIconClose
                      role="img"
                    >
                      <svg
                        role="img"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </MaterialIconClose>
                  </Icon>
                  <ForwardRef(Box)
                    display="block"
                    element="span"
                  >
                    <span>
                      articleEditor.panels.close
                    </span>
                  </ForwardRef(Box)>
                </button>
              </Styled(button)>
            </ForwardRef(NavigationButton)>
            <ForwardRef(Box)
              flexGrow={1}
            >
              <div />
            </ForwardRef(Box)>
            <Styled(span)>
              <span>
                articleEditor.panels.metadata
              </span>
            </Styled(span)>
          </div>
        </Styled(div)>
      </PanelHeader>
      <PanelSection>
        <Styled(div)
          styleProps={
            Object {
              "dark": undefined,
            }
          }
        >
          <div>
            <ForwardRef(TextInput)
              label="articleEditor.panels.preTitle"
              marginBottom={10}
              onChange={[Function]}
              value="Moonstone Uses"
            >
              <Styled(div)
                styleProps={
                  Object {
                    "marginBottom": 10,
                  }
                }
              >
                <div>
                  <Styled(label)>
                    <label>
                      <Styled(input)
                        onChange={[Function]}
                        placeholder="articleEditor.panels.preTitle"
                        styleProps={
                          Object {
                            "hasError": false,
                            "hasIcon": false,
                          }
                        }
                        value="Moonstone Uses"
                      >
                        <input
                          onChange={[Function]}
                          placeholder="articleEditor.panels.preTitle"
                          value="Moonstone Uses"
                        />
                      </Styled(input)>
                      <Styled(span)
                        styleProps={
                          Object {
                            "hasError": false,
                            "hasIcon": false,
                          }
                        }
                      >
                        <span>
                          articleEditor.panels.preTitle
                        </span>
                      </Styled(span)>
                    </label>
                  </Styled(label)>
                </div>
              </Styled(div)>
            </ForwardRef(TextInput)>
            <ForwardRef(TextInput)
              label="articleEditor.panels.title"
              marginBottom={10}
              onChange={[Function]}
              value="The Properties of Moonstone And Its Uses"
            >
              <Styled(div)
                styleProps={
                  Object {
                    "marginBottom": 10,
                  }
                }
              >
                <div>
                  <Styled(label)>
                    <label>
                      <Styled(input)
                        onChange={[Function]}
                        placeholder="articleEditor.panels.title"
                        styleProps={
                          Object {
                            "hasError": false,
                            "hasIcon": false,
                          }
                        }
                        value="The Properties of Moonstone And Its Uses"
                      >
                        <input
                          onChange={[Function]}
                          placeholder="articleEditor.panels.title"
                          value="The Properties of Moonstone And Its Uses"
                        />
                      </Styled(input)>
                      <Styled(span)
                        styleProps={
                          Object {
                            "hasError": false,
                            "hasIcon": false,
                          }
                        }
                      >
                        <span>
                          articleEditor.panels.title
                        </span>
                      </Styled(span)>
                    </label>
                  </Styled(label)>
                </div>
              </Styled(div)>
            </ForwardRef(TextInput)>
            <ForwardRef(TextInput)
              label="articleEditor.panels.lead"
              marginBottom={10}
              onChange={[Function]}
              value=""
            >
              <Styled(div)
                styleProps={
                  Object {
                    "marginBottom": 10,
                  }
                }
              >
                <div>
                  <Styled(label)>
                    <label>
                      <Styled(textarea)
                        onChange={[Function]}
                        placeholder="articleEditor.panels.lead"
                        rows={1}
                        styleProps={
                          Object {
                            "hasError": false,
                            "hasIcon": false,
                          }
                        }
                        value=""
                      >
                        <textarea
                          onChange={[Function]}
                          placeholder="articleEditor.panels.lead"
                          rows={1}
                          value=""
                        />
                      </Styled(textarea)>
                      <Styled(span)
                        styleProps={
                          Object {
                            "hasError": false,
                            "hasIcon": false,
                          }
                        }
                      >
                        <span>
                          articleEditor.panels.lead
                        </span>
                      </Styled(span)>
                    </label>
                  </Styled(label)>
                </div>
              </Styled(div)>
            </ForwardRef(TextInput)>
            <AuthorInput
              label="articleEditor.panels.authors"
              marginBottom={10}
              onChange={[Function]}
              value={
                Array [
                  Object {
                    "id": "authorId2",
                    "name": "Hermione Granger",
                  },
                ]
              }
            >
              <AutocompleteInput
                label="articleEditor.panels.authors"
                marginBottom={10}
                onChange={[Function]}
                value={
                  Array [
                    Object {
                      "id": "authorId2",
                      "name": "Hermione Granger",
                    },
                  ]
                }
                valueToChipData={[Function]}
              >
                <Downshift
                  defaultHighlightedIndex={null}
                  defaultIsOpen={false}
                  environment={[Window]}
                  getA11yStatusMessage={[Function]}
                  itemToString={[Function]}
                  onChange={[Function]}
                  onInputValueChange={[Function]}
                  onOuterClick={[Function]}
                  onSelect={[Function]}
                  onStateChange={[Function]}
                  onUserAction={[Function]}
                  scrollIntoView={[Function]}
                  selectedItem={null}
                  selectedItemChanged={[Function]}
                  stateReducer={[Function]}
                  suppressRefError={false}
                >
                  <Styled(div)
                    aria-expanded={false}
                    aria-haspopup="listbox"
                    aria-labelledby="downshift-0-label"
                    aria-owns={null}
                    role="combobox"
                    styleProps={
                      Object {
                        "marginBottom": 10,
                      }
                    }
                  >
                    <div
                      aria-expanded={false}
                      aria-haspopup="listbox"
                      aria-labelledby="downshift-0-label"
                      aria-owns={null}
                      role="combobox"
                    >
                      <Styled(label)
                        htmlFor="downshift-0-input"
                        id="downshift-0-label"
                      >
                        <label
                          htmlFor="downshift-0-input"
                          id="downshift-0-label"
                        >
                          <Chip
                            icon={[Function]}
                            key="authorId2"
                            label="Hermione Granger"
                            margin={2}
                            onIconClick={[Function]}
                          >
                            <Styled(div)
                              styleProps={
                                Object {
                                  "margin": 2,
                                }
                              }
                            >
                              <div>
                                <Styled(span)>
                                  <span>
                                    Hermione Granger
                                  </span>
                                </Styled(span)>
                                <ForwardRef(IconButton)
                                  icon={[Function]}
                                  marginLeft={-5}
                                  onClick={[Function]}
                                  variant="light"
                                >
                                  <Styled(button)
                                    onClick={[Function]}
                                    styleProps={
                                      Object {
                                        "active": undefined,
                                        "disabled": undefined,
                                        "marginLeft": -5,
                                        "variant": "light",
                                      }
                                    }
                                  >
                                    <button
                                      onClick={[Function]}
                                    >
                                      <Icon
                                        block={true}
                                        element={[Function]}
                                      >
                                        <MaterialIconClose
                                          role="img"
                                        >
                                          <svg
                                            role="img"
                                            viewBox="0 0 24 24"
                                          >
                                            <path
                                              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                            />
                                            <path
                                              d="M0 0h24v24H0z"
                                              fill="none"
                                            />
                                          </svg>
                                        </MaterialIconClose>
                                      </Icon>
                                    </button>
                                  </Styled(button)>
                                </ForwardRef(IconButton)>
                              </div>
                            </Styled(div)>
                          </Chip>
                          <Styled(input)
                            aria-activedescendant={null}
                            aria-autocomplete="list"
                            aria-controls={null}
                            aria-labelledby="downshift-0-label"
                            autoComplete="off"
                            id="downshift-0-input"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            styleProps={
                              Object {
                                "hasError": false,
                              }
                            }
                            value=""
                          >
                            <input
                              aria-activedescendant={null}
                              aria-autocomplete="list"
                              aria-controls={null}
                              aria-labelledby="downshift-0-label"
                              autoComplete="off"
                              id="downshift-0-input"
                              onBlur={[Function]}
                              onChange={[Function]}
                              onKeyDown={[Function]}
                              value=""
                            />
                          </Styled(input)>
                          <Styled(span)
                            styleProps={
                              Object {
                                "hasError": false,
                              }
                            }
                          >
                            <span>
                              articleEditor.panels.authors
                            </span>
                          </Styled(span)>
                        </label>
                      </Styled(label)>
                      <AuthorInputList
                        getItemProps={[Function]}
                        getMenuProps={[Function]}
                        highlightedIndex={null}
                        inputValue=""
                        isOpen={false}
                      >
                        <Styled(ul)
                          aria-labelledby="downshift-0-label"
                          id="downshift-0-menu"
                          role="listbox"
                        >
                          <ul
                            aria-labelledby="downshift-0-label"
                            id="downshift-0-menu"
                            role="listbox"
                          />
                        </Styled(ul)>
                      </AuthorInputList>
                    </div>
                  </Styled(div)>
                </Downshift>
              </AutocompleteInput>
            </AuthorInput>
            <TagInput
              description="articleEditor.panels.addTag"
              label="articleEditor.panels.tags"
              marginBottom={20}
              onChange={[Function]}
              value={
                Array [
                  "moonstone",
                  "potions",
                ]
              }
            >
              <Styled(div)
                styleProps={
                  Object {
                    "marginBottom": 20,
                  }
                }
              >
                <div>
                  <Styled(div)>
                    <div>
                      <Chip
                        icon={[Function]}
                        label="moonstone"
                        margin={2}
                        onIconClick={[Function]}
                      >
                        <Styled(div)
                          styleProps={
                            Object {
                              "margin": 2,
                            }
                          }
                        >
                          <div>
                            <Styled(span)>
                              <span>
                                moonstone
                              </span>
                            </Styled(span)>
                            <ForwardRef(IconButton)
                              icon={[Function]}
                              marginLeft={-5}
                              onClick={[Function]}
                              variant="light"
                            >
                              <Styled(button)
                                onClick={[Function]}
                                styleProps={
                                  Object {
                                    "active": undefined,
                                    "disabled": undefined,
                                    "marginLeft": -5,
                                    "variant": "light",
                                  }
                                }
                              >
                                <button
                                  onClick={[Function]}
                                >
                                  <Icon
                                    block={true}
                                    element={[Function]}
                                  >
                                    <MaterialIconClose
                                      role="img"
                                    >
                                      <svg
                                        role="img"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                        />
                                        <path
                                          d="M0 0h24v24H0z"
                                          fill="none"
                                        />
                                      </svg>
                                    </MaterialIconClose>
                                  </Icon>
                                </button>
                              </Styled(button)>
                            </ForwardRef(IconButton)>
                          </div>
                        </Styled(div)>
                      </Chip>
                      <Chip
                        icon={[Function]}
                        label="potions"
                        margin={2}
                        onIconClick={[Function]}
                      >
                        <Styled(div)
                          styleProps={
                            Object {
                              "margin": 2,
                            }
                          }
                        >
                          <div>
                            <Styled(span)>
                              <span>
                                potions
                              </span>
                            </Styled(span)>
                            <ForwardRef(IconButton)
                              icon={[Function]}
                              marginLeft={-5}
                              onClick={[Function]}
                              variant="light"
                            >
                              <Styled(button)
                                onClick={[Function]}
                                styleProps={
                                  Object {
                                    "active": undefined,
                                    "disabled": undefined,
                                    "marginLeft": -5,
                                    "variant": "light",
                                  }
                                }
                              >
                                <button
                                  onClick={[Function]}
                                >
                                  <Icon
                                    block={true}
                                    element={[Function]}
                                  >
                                    <MaterialIconClose
                                      role="img"
                                    >
                                      <svg
                                        role="img"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                        />
                                        <path
                                          d="M0 0h24v24H0z"
                                          fill="none"
                                        />
                                      </svg>
                                    </MaterialIconClose>
                                  </Icon>
                                </button>
                              </Styled(button)>
                            </ForwardRef(IconButton)>
                          </div>
                        </Styled(div)>
                      </Chip>
                      <Styled(input)
                        onBlur={[Function]}
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        styleProps={
                          Object {
                            "hasError": false,
                            "isPlaceholderShown": false,
                          }
                        }
                        value=""
                      >
                        <input
                          onBlur={[Function]}
                          onChange={[Function]}
                          onKeyDown={[Function]}
                          value=""
                        />
                      </Styled(input)>
                      <Styled(label)
                        styleProps={
                          Object {
                            "hasError": false,
                            "isPlaceholderShown": false,
                          }
                        }
                      >
                        <label>
                          articleEditor.panels.tags
                        </label>
                      </Styled(label)>
                    </div>
                  </Styled(div)>
                  <Styled(div)>
                    <div>
                      articleEditor.panels.addTag
                    </div>
                  </Styled(div)>
                </div>
              </Styled(div)>
            </TagInput>
            <ForwardRef(Box)>
              <div>
                <Toggle
                  checked={true}
                  label="articleEditor.panels.breakingNews"
                  onChange={[Function]}
                >
                  <label>
                    <span>
                      <span>
                        articleEditor.panels.breakingNews
                      </span>
                      <span />
                    </span>
                    <input
                      checked={true}
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span />
                  </label>
                </Toggle>
              </div>
            </ForwardRef(Box)>
          </div>
        </Styled(div)>
      </PanelSection>
      <PanelSectionHeader
        title="articleEditor.panels.image"
      >
        <Styled(div)>
          <div>
            articleEditor.panels.image
          </div>
        </Styled(div)>
      </PanelSectionHeader>
      <PanelSection
        dark={true}
      >
        <Styled(div)
          styleProps={
            Object {
              "dark": true,
            }
          }
        >
          <div>
            <ForwardRef(Box)
              height={200}
            >
              <div>
                <ForwardRef(Card)>
                  <Styled(div)
                    styleProps={Object {}}
                  >
                    <div>
                      <PlaceholderInput
                        onAddClick={[Function]}
                      >
                        <div>
                          <ForwardRef(IconButton)
                            icon={[Function]}
                            onClick={[Function]}
                            variant="large"
                          >
                            <Styled(button)
                              onClick={[Function]}
                              styleProps={
                                Object {
                                  "active": undefined,
                                  "disabled": undefined,
                                  "variant": "large",
                                }
                              }
                            >
                              <button
                                onClick={[Function]}
                              >
                                <Icon
                                  block={true}
                                  element={[Function]}
                                >
                                  <MaterialIconAdd
                                    role="img"
                                  >
                                    <svg
                                      role="img"
                                      viewBox="0 0 24 24"
                                    >
                                      <path
                                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                      />
                                      <path
                                        d="M0 0h24v24H0z"
                                        fill="none"
                                      />
                                    </svg>
                                  </MaterialIconAdd>
                                </Icon>
                              </button>
                            </Styled(button)>
                          </ForwardRef(IconButton)>
                        </div>
                      </PlaceholderInput>
                    </div>
                  </Styled(div)>
                </ForwardRef(Card)>
              </div>
            </ForwardRef(Box)>
          </div>
        </Styled(div)>
      </PanelSection>
      <PanelSectionHeader
        title="articleEditor.panels.peering"
      >
        <Styled(div)>
          <div>
            articleEditor.panels.peering
          </div>
        </Styled(div)>
      </PanelSectionHeader>
      <PanelSection>
        <Styled(div)
          styleProps={
            Object {
              "dark": undefined,
            }
          }
        >
          <div>
            <Toggle
              checked={false}
              description="articleEditor.panels.allowPeerPublishing"
              label="articleEditor.panels.peerShare"
              onChange={[Function]}
            >
              <label>
                <span>
                  <span>
                    articleEditor.panels.peerShare
                  </span>
                  <span>
                    articleEditor.panels.allowPeerPublishing
                  </span>
                </span>
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <span />
              </label>
            </Toggle>
          </div>
        </Styled(div)>
      </PanelSection>
    </div>
  </Styled(div)>
  <Drawer
    open={false}
    width={480}
  >
    <Modal
      open={false}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={200}
        unmountOnExit={true}
      />
    </Modal>
  </Drawer>
  <Drawer
    open={false}
    width={480}
  >
    <Modal
      open={false}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={200}
        unmountOnExit={true}
      />
    </Modal>
  </Drawer>
</ArticleMetadataPanel>
`;
